src/ProjectSystem/Impl/IO/MsBuildFileSystemWatcherEntries.cs

```diff
@@ -16,6 +16,16 @@ namespace Microsoft.VisualStudio.ProjectSystem.FileSystemMirroring.IO {
     public class MsBuildFileSystemWatcherEntries {
         private readonly Dictionary<string, Entry> _entries = new Dictionary<string, Entry>(StringComparer.OrdinalIgnoreCase);
 
+        public bool ContainsFileEntry(string relativeFilePath) {
+            return _entries.ContainsKey(relativeFilePath);
+        }
+
+        public bool ContainsDirectoryEntry(string relativePath) {
+            relativePath = PathHelper.EnsureTrailingSlash(relativePath);
+            return _entries.Values.Where(v => v.RelativePath.StartsWithIgnoreCase(relativePath) ||
```

Comment:
`_entries.Values.Any(v => v.RelativePath.StartsWithIgnoreCase(previousRelativePath) ||
                                v.ShortPath.StartsWithIgnoreCase(previousRelativePath))`
And we also can extract this condition into method - it is used in many places